### PR TITLE
Upgrade Mesh to 0x Protocol V3

### PR DESCRIPTION
**Pull request by fabioberger**
This PR upgrades Mesh to work with 0x Protocol V3. 

**WARNING:** This is opened against `refactor/orderEvents`. Once that lands, it should be re-opened against development.

## Minor complication

The protocol team used a slightly different domain struct then the [standard specifies](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md#definition-of-domainseparator), using `verifyingContractAddress` instead of `verifyingContract`. This is [being fixed](https://github.com/0xProject/0x-monorepo/pull/2172). Until then, you need to find-replace `verifyingContract` with `verifyingContractAddress` in `go-ethereum/signer/core/signed_data.go` in your project's vendor dir to get things to work. Once #2172 lands, we will need to do the following:
- [x] Revert changes made to `go-ethereum/signer/core/signed_data.go`
- [x] Generate a new Ganache snapshot with the updated contracts
- [x] ~~Update the hard-coded `DomainHash`~~
- [x] Update `VerifyingContractAddress` to `VerifyingContract` in Mesh codebase
- [x] Update the hard-coded test fixture order hash and signature in `order_test.go`

## Outstanding TODOs
- [x] ~~DevUtils order validation `eth_call` fails with result '0x'. I believe this is due to the [Ganache snapshot](http://ganache-snapshots.0x.org.s3.amazonaws.com/0x_ganache_snapshot-v3-beta-4.2.0.zip) we're using having been generated using an old version of Ganache with a bug. We need to get Jacob to generate a new snapshot and give it a try.~~ @fabioberger I can't reproduce this issue. Maybe it was fixed?
- [x] Update 0xorg/ganache-cli docker image referenced in CircleCI config to V3 snapshot. 
- [x] `assert.Equal` failing despite no diff in event_decoder_test.go line 183. Fix this.
- [x] Add a database index for `MakerFeeAssetData`.
- [x] Add `MakerFeeAssetData` to `EventDecoder` when watching new orders.
- [x] The introduction of `MakerFeeAssetData` means that there is now one more field that could impact an orders validity. In `OrderWatcher`, we should also re-validate orders when the asset encoded in`MakerFeeAssetData` is transferred or it's allowance is updated. This can be done on lines 289, 309, 329 and 349.
- [x] The `@0x/mesh-browser` package and browser integration tests need to be updated once we've published the TS libraries V3 release-candidates.
- [x] ~~Try and get rid of the hack with description: `// HACK(fabio): There is a bug in Ganache-core where eth_call fails without specifying...`. This is probably also due to the old Ganache version used to generate the snapshot. That is my hunch at least.~~ @fabioberger I couldn't find this comment anywhere. Maybe you forgot to push the latest commit? @albrow yep, forgot to push those comments. No worries.

### CODE REVIEW

**Comment by albrow** [comment]
This PR is currently blocked by the following:

1. The `0xorg/ganache-cli` Docker image which we are using for tests is crashing with the error message:

```
Error: nonce generation function failed or private key is invalid
    at Object.t.sign (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:52:264461)
    at Object.sign (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:52:253479)
    at Object.t.ecsign (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:17296)
    at e.sign (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:88084)
    at e.sign (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:84960)
    at e.hash (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:61:9536)
    at e.d [as hash] (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:82139)
    at o.combine (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:81172)
    at new o (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:80848)
    at Function.o.fromResults (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:10:81931)
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:47:121203
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:61:1833282
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:26124
    at i (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:41179)
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:61:1190532
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:105418
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:32:392
    at c (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:32:5407)
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:32:317
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:61:1851708
    at /usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:23237
    at o (/usr/lib/node_modules/ganache-cli/build/ganache-core.node.cli.js:2:26646)
```

2. We haven't finalized/standardized the JSON format for v3 orders.

3. We haven't published any v3 packages to NPM. We need these for the integration tests and browser example.

It's possible that other blockers might come up after these are addressed, but this is everything that I know of.

**Comment by albrow** [comment]
I'm assuming https://github.com/0xProject/0x-monorepo/pull/2212 is the final state for order schemas so I updated this PR to match. This means removing the `DomainHash` field and adding `ChainID` and `ExchangeAddress`.

File: scenario/scenario.go [hunk 1]
```diff
@@ -22,19 +22,23 @@ import (
...
 	testOrder := &zeroex.Order{
-		MakerAddress:          makerAddress,
-		TakerAddress:          constants.NullAddress,
-		SenderAddress:         constants.NullAddress,
+		ChainID:         big.NewInt(constants.TestNetworkID),
+		ExchangeAddress: ethereum.NetworkIDToContractAddresses[constants.TestNetworkID].Exchange,
+		MakerAddress:    makerAddress,
+		TakerAddress:    constants.NullAddress,
+		SenderAddress:   constants.NullAddress,
+		// TODO(albrow): We should remove MakerFeeAssetData and TakerFeeAssetData after the DevUtils contract is fixed
```
Comment (albrow) [inline scenario/scenario.go]: @fabioberger has the Ganache snapshot been updated to use the latest `DevUtils` contract? If so we should set the `MakerFeeAssetData` and `TakerFeeAssetData` to `NullBytes` and remove this `TODO`. I made this comment in a few other places as well.

File: zeroex/ordervalidator/order_validator.go [hunk 2]
```diff
@@ -335,6 +340,12 @@ func (o *OrderValidator) BatchValidate(rawSignedOrders []*zeroex.SignedOrder, ar
 				ctx, cancel := context.WithTimeout(context.Background(), getOrderRelevantStateTimeout)
 				defer cancel()
 				opts := &bind.CallOpts{
+					// HACK(albrow): From field should not be required for eth_call but
+					// including it here is a workaround for a bug in Ganache. Removing
+					// this line causes Ganache to crash.
+					// TODO(albrow): Test if this causes problems for real Ethereum nodes
```
Comment (albrow) [inline zeroex/ordervalidator/order_validator.go]: @fabioberger I think you mentioned this shouldn't cause any problems for real Ethereum nodes. If that's the case we should remove this `TODO` (but keep the `HACK` comment).

File: zeroex/orderwatch/order_watcher.go [hunk 3]
```diff
@@ -583,15 +589,28 @@ func (w *Watcher) findOrder(orderHash common.Hash) *meshdb.Order {
...
+// tokenID.
+func (w *Watcher) findOrdersByTokenAddressAndTokenID(makerAddress, tokenAddress common.Address, tokenID *big.Int) ([]*meshdb.Order, error) {
+	ordersWithAffectedMakerAsset, err := w.meshDB.FindOrdersByMakerAddressTokenAddressAndTokenID(makerAddress, tokenAddress, tokenID)
+	if err != nil {
+		logger.WithFields(logger.Fields{
+			"error": err.Error(),
+		}).Error("unexpected query error encountered")
+		return nil, err
+	}
+	// TODO(albrow): Test that orders with a relevant MakerFeeAssetData are
```
Comment (albrow) [inline zeroex/orderwatch/order_watcher.go]: I did add this test so we can now remove this `TODO`.

File: scenario/scenario.go (hunk 1 above)
Comment (fabioberger) [reply scenario/scenario.go]: Fixed!

**Comment by fabioberger** [comment]
Accidentally merged this too soon. I've reverted the changes on `development`.

**Pull request merged**